FirebaseAuth/Sources/User/FIRUser.m

```diff
@@ -1098,15 +1104,107 @@ - (void)linkWithCredential:(FIRAuthCredential *)credential
       FIREmailPasswordAuthCredential *emailPasswordCredential =
           (FIREmailPasswordAuthCredential *)credential;
       if (emailPasswordCredential.password) {
-        [self updateEmail:emailPasswordCredential.email
-                 password:emailPasswordCredential.password
-                 callback:^(NSError *error) {
-                   if (error) {
-                     callInMainThreadWithAuthDataResultAndError(completion, nil, error);
-                   } else {
-                     callInMainThreadWithAuthDataResultAndError(completion, result, nil);
-                   }
-                 }];
+        [self internalGetTokenWithCallback:^(NSString *_Nullable accessToken,
```

Comment:
Can you explain a bit why so much code has to change here? I thought we only needed to call a different backend API?

Reply:
Sure. Basically `updateEmail` is used in other places (like profile change) as well, so we can't just update its underlying backend API. Here, I reimplement the full flow to switch to`signUp` endpoint.

Reply:
Can we pull this into an internal "linkWithEmailCredential" method? Or can we reuse the signUp method in https://github.com/firebase/firebase-ios-sdk/blob/0c8d1433b84e0c81e19f4987568c7b80027111a9/FirebaseAuth/Sources/Auth/FIRAuth.m#L1990 to pass the new callback and make that method accept idToken as an optional param?

Since this method already has several cases (emailAuthCredential (with and without password), other types of credentials), i think making/reusing a helper method would be great.

Reply:
Pulled in to a helper method. Gonna create a helper method to handle the recaptcha related flows in a separate pr like what we plan to do in web.

It's hard to reuse the signUp method since linking is a user method while sign up is an auth method so there are many differences .

Reply:
I was able to use a helper method for most of the recaptcha flows in Swift in https://github.com/firebase/firebase-ios-sdk/pull/11942/files#diff-df7cc3b7882f0d0525d338279d734310f97c320a9af8e86265d2d7430b8086fa. It will be trickier in ObjC, but some of the ideas might apply.

Reply:
Cool, thanks Paul!